includes/sanitizers/class-amp-bento-sanitizer.php

```diff
@@ -0,0 +1,264 @@
+			) {
+				$script->parentNode->removeChild( $script );
+			} else {
+				ValidationExemption::mark_node_as_px_verified( $script );
+				$non_amp_scripts_retained++;
+			}
+		}
+
+		// Remove the Bento runtime script if it is not needed, or else mark it as PX-verified.
+		$bento_runtime_scripts = $this->dom->xpath->query(
+			'//script[ @src = "https://cdn.ampproject.org/bento.mjs" or @src = "https://cdn.ampproject.org/bento.js" ]'
+		);
+		if ( 0 === $non_amp_scripts_retained ) {
+			foreach ( $bento_runtime_scripts as $bento_runtime_script ) {
+				$bento_runtime_script->parentNode->removeChild( $bento_runtime_script );
+			}
+		} else {
+			foreach ( $bento_runtime_scripts as $bento_runtime_script ) {
+				ValidationExemption::mark_node_as_px_verified( $bento_runtime_script );
+			}
+		}
+
+		// If bento-prefixed components were discovered, then ensure that the tag-and-attribute sanitizer will prefer
+		// Bento components when validating and that it will use the Bento versions of component scripts, and ultimately
+		// AMP_Theme_Support::ensure_required_markup() will add the Bento experiment opt-in which is still required at
+		// the moment.
+		if ( count( $bento_elements_discovered ) > 0 && $this->tag_and_attribute_sanitizer ) {
+			$this->tag_and_attribute_sanitizer->update_args(
+				[ 'prefer_bento' => true ]
+			);
+		}
+	}
+
+	/**
+	 * Adapt inline styles from Bento element to AMP layout attributes.
+	 *
+	 * This will try its best to convert `width`, `height`, and `aspect-ratio` inline styles over to their corresponding
+	 * AMP layout attributes. In order for a Bento component to be AMP-compatible, it needs to utilize inline styles
+	 * for its dimensions rather than rely on a stylesheet rule.
+	 *
+	 * @param Element $amp_element AMP element (converted from Bento).
+	 */
+	private function adapt_layout_styles( Element $amp_element ) {
+		$style_string = $amp_element->getAttribute( Attribute::STYLE );
+		if ( ! $style_string ) {
+			return;
+		}
+
+		$styles = $this->parse_style_string( $style_string );
+
+		$layout_attributes = [];
+
+		if ( isset( $styles['height'] ) ) {
+			$height = new CssLength( $styles['height'] );
+			$height->validate( false, false );
+			if ( $height->isValid() ) {
+				$layout_attributes[ Attribute::HEIGHT ] = $height->getNumeral() . ( $height->getUnit() !== 'px' ? $height->getUnit() : '' );
+				unset( $styles['height'] );
+			}
+		}
+
+		if ( ! isset( $styles['width'] ) || '100%' === $styles['width'] ) {
+			$layout_attributes[ Attribute::WIDTH ]  = 'auto';
+			$layout_attributes[ Attribute::LAYOUT ] = Layout::FIXED_HEIGHT;
+			unset( $styles['width'] );
+		} else {
+			$height = new CssLength( $styles['width'] );
+			$height->validate( false, false );
+			if ( $height->isValid() ) {
+				$layout_attributes[ Attribute::WIDTH ] = $height->getNumeral() . ( $height->getUnit() !== 'px' ? $height->getUnit() : '' );
+				unset( $styles['width'] );
+			}
```

Comment:
Should the variable be named `$width` instead of `$height`?

Reply:
Ooof, yes, I think you're right. Copypasta my bad.

```suggestion
			$width = new CssLength( $styles['width'] );
			$width->validate( false, false );
			if ( $width->isValid() ) {
				$layout_attributes[ Attribute::WIDTH ] = $width->getNumeral() . ( $width->getUnit() !== 'px' ? $width->getUnit() : '' );
				unset( $styles['width'] );
			}
```